source/mqtt3_to_mqtt5_adapter.c

```diff
@@ -260,8 +269,451 @@ static int s_aws_mqtt5_adapter_perform_safe_callback(
+             * error code of AWS_ERROR_MQTT_CONNECTION_RESET_FOR_ADAPTER_CONNECT) we only react to connection failures
+             * if
+             *   (1) the error code is not AWS_ERROR_MQTT_CONNECTION_RESET_FOR_ADAPTER_CONNECT and
+             *   (2) we're in the FIRST_CONNECT state
+             *
+             * Only if both of these are true should we invoke the connection completion callback with a failure and
+             * put the adapter into the "disconnected" state, simulating the way the 311 client stops after an
+             * initial connection failure.
+             */
+            if (event->error_code != AWS_ERROR_MQTT_CONNECTION_RESET_FOR_ADAPTER_CONNECT &&
+                adapter->adapter_state == AWS_MQTT_AS_FIRST_CONNECT) {
+
+                if (adapter->on_connection_complete != NULL) {
+                    (*adapter->on_connection_complete)(
+                        &adapter->base, event->error_code, 0, false, adapter->on_connection_complete_user_data);
+
+                    adapter->on_connection_complete = NULL;
+                    adapter->on_connection_complete_user_data = NULL;
+                }
+
+                adapter->adapter_state = AWS_MQTT_AS_STAY_DISCONNECTED;
+            }
+            break;
+
+        case AWS_MQTT5_CLET_DISCONNECTION:
+            /*
+             * If the 311 view is that we're in the stay-connected state (ie we've successfully done or simulated
+             * an initial connection), then invoke the connection interrupted callback.
+             */
+            if (adapter->on_interrupted != NULL && adapter->adapter_state == AWS_MQTT_AS_STAY_CONNECTED &&
+                event->error_code != AWS_ERROR_MQTT_CONNECTION_RESET_FOR_ADAPTER_CONNECT) {
+
+                (*adapter->on_interrupted)(&adapter->base, event->error_code, adapter->on_interrupted_user_data);
+            }
+            break;
+
+        case AWS_MQTT5_CLET_STOPPED:
+            /* If an MQTT311-view user is waiting on a disconnect callback, invoke it */
+            if (adapter->on_disconnect) {
+                (*adapter->on_disconnect)(&adapter->base, adapter->on_disconnect_user_data);
+
+                adapter->on_disconnect = NULL;
+                adapter->on_disconnect_user_data = NULL;
+            }
+
+            if (adapter->on_closed) {
+                (*adapter->on_closed)(&adapter->base, NULL, adapter->on_closed_user_data);
+            }
+
+            /*
+             * Judgement call: If the mqtt5 client is stopped behind our back, it seems better to transition to the
+             * disconnected state (which only requires a connect() to restart) then stay in the STAY_CONNECTED state
+             * which currently requires a disconnect() and then a connect() to restore connectivity.
+             *
+             * ToDo: what if we disabled mqtt5 client start/stop somehow while the adapter is attached, preventing
```

Comment:
If we take this path, both start() and stop() for mqtt5 return an int that's currently just whether or not s_aws_mqtt5_client_change_desired_state() was successfully scheduled. Could we add a check in that function or start() and stop() itself and return/log a new error that clearly states that start() and/or stop() aren't available when an adapter is attached and not allow the user to call them?

Reply:
After recent discussion, this doesn't make much sense since the mqtt5 client could be being used as a straight up mqtt5 client.